proteus/WaveTools.pyx

```diff
@@ -0,0 +1,1355 @@
+# A type of -*- python -*- file
+#cython: embedsignature=True
+"""Tools for working with water waves.
+
+The primary objective of this module is to provide solutions (exact and
+approximate) for the free surface deformation and subsurface velocity
+components of water waves. These can be used as boundary conditions, wave
+generation sources, and validation solutions for numerical wave codes.
+
+.. inheritance-diagram:: ShockCapturing
+    :parts: 2
+"""
+from math import pi, tanh, sqrt, exp, log, sin, cos, cosh, sinh
+import numpy as np
+import cmath as cmath
+from Profiling import logEvent
+import time as tt
+import sys as sys
+
+
+def setVertDir(g):
+    return -g/(sqrt(g[0]**2 + g[1]**2 + g[2]**2))
+
+def setDirVector(vector):
+    return vector/(sqrt(vector[0]**2 + vector[1]**2 + vector[2]**2))
+
+def dirCheck(v1, v2):
+    dircheck = abs(v1[0]*v2[0]+v1[1]*v2[1]+v1[2]*v2[2])
+        #print self.dircheck
+    if dircheck > 1e-10:
+        logEvent("Wave direction is not perpendicular to gravity vector. Check input",level=0)
+        return sys.exit(1)
+    else:
+        return None
+def reduceToIntervals(fi,df):
+    fim_tmp = (0.5*(fi[1:]+fi[:-1])).tolist()
+    return np.array([fim_tmp[0]-0.5*df]+fim_tmp+[fim_tmp[-1]+0.5*df])
+def returnRectangles(a,x):
+    return 0.5*(a[1:]+a[:-1])*(x[1:]-x[:-1])
+def returnRectangles3D(a,x,y):
+
+    ai = 0.5*(a[1:,:]+a[:-1,:])
+    ai = 0.5*(ai[:,1:]+ai[:,:-1])
+    for ii in range(len(x)-1):
+        ai[ii,:] *= (y[1:]-y[:-1]) 
+    for jj in range(len(y) - 1):
+        ai[:,jj] *= (x[1:]-x[:-1]) 
+        
+    
+    return ai
+def normIntegral(Sint,th):
+    G0 = 1./sum(returnRectangles(Sint,th))
+    return G0*Sint
+
+
+
+def eta_mode(x,y,z,t,kDir,omega,phi,amplitude):
+    """Returns a single frequency mode for free-surface elevation at point x,y,z,t
+    :param kDir: wave number vector [1/L]
+    :param omega: angular frequency [1/T]
+    :param phi: phase [0,2*pi]
+    :param amplitude: wave amplitude [L/T^2]
+    """
+    phase = x*kDir[0]+y*kDir[1]+z*kDir[2] - omega*t  + phi
+    return amplitude*cos(phase)
+
+
+def vel_mode(x,y,z,t,kDir,kAbs,omega,phi,amplitude,mwl,depth,g,vDir,comp):
+    """Returns a single frequency mode for velocity at point x,y,z,t
+    :param kDir: wave number vector [1/L]
+    :param omega: angular frequency [1/T]
+    :param phi: phase [0,2*pi]
+    :param amplitude: wave amplitude [L/T^2]
+    :param mwl: mean water level [L]
+    :param depth: water depth [L]
+    :param g: gravity vector
+    :param vDir (vertical direction - opposite to the gravity vector)
+    :param comp: component "x", "y" or "z"
+    """
+
+    phase = x*kDir[0]+y*kDir[1]+z*kDir[2] - omega*t  + phi
+    Z =  (vDir[0]*x + vDir[1]*y+ vDir[2]*z) - mwl
+# We should not really be enforcing the velocity outside these bounds
+    if( (Z < -depth) or (Z > 2*amplitude)):
```

Comment:
I think that this implementation makes it really hard (and impossible in some cases) to set boundary conditions. If you have a wave series with variable wave heights, the region in space that will raise this error changes for each amplitude. I can't just mark a section of the boundary and set a Dirichlet BC on that boundary.